### ISSUE TITLE
Custom Loss from External Optimisation Function? 

### ISSUE THREAD

**Issue by jordan-bird** [pytorch/pytorch#18200]
Hello,

Sorry if this is the wrong place to open an issue like this but I can't find any related work, and StackOverflow has not been able to help (1 week stale, 0 replies)

I am trying to replace the loss of my network that has 3 outputs by an external function

My code is as follows:

```
def my_loss(output, target):
    global var
    table = str.maketrans(dict.fromkeys('tensor()'))
    ftn = fitness(str(output.data[0][0]).translate(table) + ", " + str(output.data[0][1]).translate(table) + ", " + str(output.data[0][2]).translate(table))

    test = -var
	
	# I just want 'ftn' to be the loss of the network

    #loss = torch.mean(output[0][1]+ftn)
	
    loss = torch.tensor([ftn], requires_grad=True)
    
    return loss
```

ftn is a variable that calculates a score of the three outputs by passing their outputs to fitness(x,y,z) where the fitness is between 0 and 100, where 0 is the best and 100 is the worst (perfect for loss reduction). fitness() is way too long to post here but basically it's a problem that optimisation algorithms (genetic, PSO etc) are **very** bad at solving, so I want to explore the NN approach just in case it may be decent, or at least comparably bad. 

Basically, I want to completely ignore the data, since it's all dummy data. I want the network to learn to produce the three inputs to the function (x, y, z) where they are the values of the three final neurons. This is why I don't consider it in the loss. 

This must be possible, and I think I'm only a line of code or two away. 

When I run this, the first random output is evaluated at epoch 0, and then following this, nothing changes. The network outputs the same, the function evaluates the same, and the loss is always the same. No learning occurs.

It would be really interesting to see a solution to this even if it's hacky (can't get hackier than mine, right?), because this is nigh on impossible to implement within TensorFlow graphs but in PyTorch Python logic WILL run every epoch within loss. 

Outputs are rounded since f() only accepts whole numbers. I set the learning rate to a high value of 0.5 to account for this but I will tweak when it's working properly. Same goes for momentum.

Thank you!

tl;dr: Ignore all input data. f(x,y,z) will output a value between 0.0 to 100.0 where 100.0 is the worst. x,y,z are the outputs of the final layer. The goal is to minimise f(x,y,z) by learning good combinations of x,y,z. The results of f(x1,y1,z1) should influence f(x2,y2,z3) after backpropagating the loss which is the findings of f().

(A solution would be **amazing** and I will buy you a beer as a thank you)

PS: My full code is as follows:

```
import torch
import torch.nn as nn

from testnetwork import *

var = 2

n_in, n_h, n_out, batch_size = 1, 5, 3, 1

x = torch.tensor([[1.0]])
y = torch.tensor([[1.0], [1.0], [1.0], [1.0], [1.0], [1.0], [1.0], [0.0], [1.0], [1.0]])

model = nn.Sequential(nn.Linear(n_in, n_h),
                     nn.ReLU(),
                     nn.ReLU()
                     )

def fitness(string):
    print(string)
    list = string.split(",")
    x = (int(round(float(list[0]))))
    y = (int(round(float(list[1]))))
    z = (int(round(float(list[2]))))
    print(list)
    loss = 100 - testnetwork(x, y, z)
    return loss


def my_loss(output, target):
    global var
    table = str.maketrans(dict.fromkeys('tensor()'))
    ftn = fitness(str(output.data[0][0]).translate(table) + ", " + str(output.data[0][1]).translate(table) + ", " + str(output.data[0][2]).translate(table))

    test = -var
	
	# I just want 'ftn' to be the loss of the network

    #loss = torch.mean(output[0][1]+ftn)
	
    loss = torch.tensor([ftn], requires_grad=True)
    
    return loss



optimizer = torch.optim.SGD(model.parameters(), lr=0.5)
#optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

for epoch in range(10):
    # Forward Propagation
    y_pred = model(x)
    # Compute and print loss
    loss = my_loss(y_pred, y)
    print('epoch: ', epoch,' loss: ', loss.item())
    # Zero the gradients
    #optimizer.zero_grad()
    
    # perform a backward pass (backpropagation)
    loss.backward(retain_graph=True)
    
    # Update the parameters
    optimizer.step()
```

EDIT: PPS: the function is testnetwork(x, y, z) - where x y and z are the outputs

**Comment by soumith** [comment]
please use https://discuss.pytorch.org for questions

StackOverflow is not our primary source of support (afaik no one handles those questions).
